rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,881 @@
+  jwt:
+    jwt:
+      algorithm: ECDSA_P256_SHA256
+      allowed_subject_algorithms: [ECDSA_P256_SHA256]
+  saml_idp:
+    tls:
+      algorithm: ECDSA_P256_SHA256
+      allowed_subject_algorithms: [ECDSA_P256_SHA256]
+  oidc_idp:
+    jwt:
+      algorithm: ECDSA_P256_SHA256
+      allowed_subject_algorithms: [ECDSA_P256_SHA256]
+```
+
+For backward-compatibility, all certs already signed by trusted CAs will
+continue to be trusted, `allowed_subject_algorithms` can be modified at any time
+without breaking connectivity, and only controls the allowed algorithms used for
+new certificates signed by the CA.
+
+Changing CA `algorithm` values in this config will take effect for:
+
+* new Teleport clusters
+* existing Teleport clusters only after a CA rotation.
+
+### Algorithms
+
+These algorithms are being considered for support:
+
+#### RSA
+
+Private key sizes: 2048, 3072, 4096
+
+Signature algorithms: PKCS#1 v1.5
+
+Digest/hash algorithms: SHA512 for SSH, SHA256 for TLS
+
+Considerations:
+
+* RSA2048 is the current default and deviating from it by default may break
+  compatibility with third-party components and protocols
+* RSA has the most widespread support among all protocols
+* Certain database protocols only support RSA client certs
+  * <https://docs.snowflake.com/en/user-guide/key-pair-auth#step-2-generate-a-public-key>
+* Some apps may only support RSA signed JWTs
+* If we must continue to support RSA, we might as well support larger key sizes
+  (at least for CA keys), 3072 and 4096-bit are the most commonly used and
+  supported by e.g. GCP KMS.
+* golang.org/x/crypto/ssh uses SHA512 hash by default with all RSA public keys
+  (but this can be overridden)
+  <https://github.com/golang/crypto/blob/0ff60057bbafb685e9f9a97af5261f484f8283d1/ssh/certs.go#L443-L445>
+* crypto/x509 uses SHA256 hash by default with all RSA public keys
+  (but this can be overridden)
+  <https://github.com/golang/go/blob/dbf9bf2c39116f1330002ebba8f8870b96645d87/src/crypto/x509/x509.go#L1411-L1414>
+* ssh only supports the PKCS#1 v1.5 signature scheme with RSA keys
+  <https://datatracker.ietf.org/doc/html/rfc8332>
+* FIPS 186-5 approves all listed options
+* BoringCrypto supports all listed options
+* We could consider PSS signatures instead of PKCS#1 v1.5 for TLS and JWT
+  signatures, but SSH does not support it.
+
+#### ECDSA
+
+Curves: P-256
+
+Digest/hash algorithms: SHA256
+
+Considerations:
+
+* ECDSA has good support across SSH and TLS protocols for both client and CA
+  certs.
+* ECDSA certs are supported by web browsers.
+* ECDSA key generation is *much* faster than RSA key generation.
+* ECDSA signatures are faster than RSA signatures.
+* FIPS 186-5 approves all listed options
+* BoringCrypto supports all listed options
+* The P-256 curve is the most popular, it is considered to be secure, and it has
+  the broadest support among external tools.
+* We could consider supporting the P-384 and P-521 curves for CAs.
```

Comment:
I'm a bit concerned that people are going to blindly start using the highest number that they see even when the drawbacks offset the advantages; P-384 could make sense for a CA (although 256-bit ecdsa is roughly equivalent to 3072 bit RSA already, which is more than what we're currently defaulting to), P-521 is almost guaranteed to be excessive.

Reply:
Also of note is that both signatures and verifications are about 10 times slower with P-384 than with P-256, with P-521 being roughly 2.5 to 3 times slower than P-384 (as per `go test -bench="(BenchmarkSign|BenchmarkVerify)" crypto/rsa crypto/ecdsa`).

Reply:
just to be clear, I'm not recommending we support those, they are not listed in the CA sections below